docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+* BenchmarkDotNet runs every benchmark in a separate process, process isolation allows avoiding side-effects. The more memory allocated by given benchmark, the bigger the difference for in-proc vs out-proc execution.
+* BenchmarkDotNet was designed to make accurate nano-benchmarks with repeatable results possible, to achieve that it does many things, including overhead calculation and subtraction (it benchmarks an empty method with the same signature and subtract the average value from results).
+* BenchmarkDotNet removes outliers by default (this repo is configured to remove only the upper outliers)
+* BenchmarkDotNet creates a type which derives from type with benchmarks. So the type with benchmarks must **not** be **sealed** and it can **not** be **static** and it has to be **public**. It also has to be a `class` (no structs support).
+
+## Setup
+
+Let's write a benchmark that measures the performance of reverting an array of 1 000 integers:
+
+```cs
+[Benchmark]
+public void Reverse()
+{
+    int[] array = Enumerable.Range(0, 1_000).ToArray();
+
+    Array.Reverse(array);
+}
+```
+
+Profile it using the [ETW Profiler](./benchmarkdotnet.md#Profiling):
+
+```cmd
+dotnet run -c Release -f netcoreapp3.0 --filter *.Reverse --profiler ETW
+```
+
+And open the produced trace file with [PerfView](https://github.com/Microsoft/perfview):
+
+![Regression](img/setup_array_reverse_profiler.png)
+
+As you can see, reversing the array took only `26.6%` of the benchmark execution time!! The rest was spent on executing the array creation logic. What does it mean? It means that the given benchmark is measuring the performance of creating and reversing the array. Not just reversing the array.
+
+This is why **the initialization logic should be always separated from the benchmark**. If possible of course.
+
+### GlobalSetup
+
+Public method marked with `[GlobalSetup]` attribute is going to be executed **exactly once**, before running the benchmark for the first time.
+
+```cs
+private int[] _array;
+
+[GlobalSetup]
+public void SetupReverse() => _array = Enumerable.Range(0, 1000).ToArray();
+
+[Benchmark]
+public void Reverse() => Array.Reverse(_array);
+```
+
+If given `[GlobalSetup]` does not specify the `Target(s)` benchmark name(s), it's executed for every benchmark declared in a given class.
+
+In case you want to have a class with multiple benchmarks and multiple, dedicated setup methods you need to use `Target` or `Targets` properties of the setup attribute:
+
+```cs
+[GlobalSetup(Target = nameof(Reverse))]
+public void SetupReverse()
+
+[GlobalSetup(Targets = new [] { nameof(Array), nameof(Span), nameof(ReadOnlySpan)})]
+public void SetupArray()
+```
+
+**Note:** If you need to clean up resources after the benchmark run, you should use the corresponding `[GlobalCleanup]` attribute. It's going to be executed only once, after all benchmark iterations.
```

Comment:
Worth it to call out here that an example of required cleanup would be any files created on the disk by the benchmark process.